### ISSUE TITLE
"Built on" date always set to 1970-01-01

### ISSUE THREAD

**Issue by adtya** [nix-community/lanzaboote#136]
In both title and version for the boot entry, the "Built on" date is always set to 1970-01-01. 

The below output is taken from `bootctl status`.
```
Default Boot Loader Entry:
         type: Boot Loader Specification Type #2 (.efi)
        title: NixOS Stoat 23.05.20230315.ac718d0 (Linux 6.2.6-zen1) (Generation 79, Built on 1970-01-01)
           id: nixos-generation-79.efi
       source: /boot/EFI/Linux/nixos-generation-79.efi
     sort-key: lanza
      version: Generation 79, Built on 1970-01-01
        linux: EFI/Linux/nixos-generation-79.efi
      options: init=/nix/store/apv56ilylfas2ilpcrjhg0116nk0w08x-nixos-system-Skipper-23.05.20230315.ac718d0/init
```

could the `fs::metadata(path)?.mtime()` in `read_build_time` return 0 for any reason?
I've checked the mtime on the generations files and they all have the correct mtime. 
```
$ stat /nix/var/nix/profiles/system-79-link 
  File: /nix/var/nix/profiles/system-79-link -> /nix/store/apv56ilylfas2ilpcrjhg0116nk0w08x-nixos-system-Skipper-23.05.20230315.ac718d0
  Size: 87              Blocks: 8          IO Block: 4096   symbolic link
Device: 0,36    Inode: 5434462     Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-03-19 13:22:53.467218885 +0530
Modify: 2023-03-19 13:22:53.400212392 +0530
Change: 2023-03-19 13:22:53.400212392 +0530
 Birth: 2023-03-19 13:22:53.400212392 +0530
 ```

**Comment by SuperSandro2000** [comment]
> could the `fs::metadata(path)?.mtime()` in `read_build_time` return 0 for any reason?

If it is run against a file in the nix store, this is expected.

**Comment by adtya** [comment]
files in nix store doesn't seem to have an mtime,
```
❯ stat /nix/store/apv56ilylfas2ilpcrjhg0116nk0w08x-nixos-system-Skipper-23.05.20230315.ac718d0                                                                              
  File: /nix/store/apv56ilylfas2ilpcrjhg0116nk0w08x-nixos-system-Skipper-23.05.20230315.ac718d0                                                                             
  Size: 418             Blocks: 0          IO Block: 4096   directory                                                                                                       
Device: 0,36    Inode: 5434435     Links: 1                                                                                                                                 
Access: (0555/dr-xr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)                                                                                                    
Access: 2023-03-19 13:22:53.000000000 +0530                                                                                                                                 
Modify: 1970-01-01 05:30:01.000000000 +0530                                                                                                                                 
Change: 2023-03-19 13:22:53.240196887 +0530                                                                                                                                 
 Birth: 2023-03-19 13:22:53.088182157 +0530
```

In that case, wouldn't it make sense to use `ctime` instead of `mtime`?

**Comment by RaitoBezarius** [comment]
> files in nix store doesn't seem to have an mtime,
> 
> ```
> ❯ stat /nix/store/apv56ilylfas2ilpcrjhg0116nk0w08x-nixos-system-Skipper-23.05.20230315.ac718d0                                                                              
>   File: /nix/store/apv56ilylfas2ilpcrjhg0116nk0w08x-nixos-system-Skipper-23.05.20230315.ac718d0                                                                             
>   Size: 418             Blocks: 0          IO Block: 4096   directory                                                                                                       
> Device: 0,36    Inode: 5434435     Links: 1                                                                                                                                 
> Access: (0555/dr-xr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)                                                                                                    
> Access: 2023-03-19 13:22:53.000000000 +0530                                                                                                                                 
> Modify: 1970-01-01 05:30:01.000000000 +0530                                                                                                                                 
> Change: 2023-03-19 13:22:53.240196887 +0530                                                                                                                                 
>  Birth: 2023-03-19 13:22:53.088182157 +0530
> ```
> 
> In that case, wouldn't it make sense to use `ctime` instead of `mtime`?

Hm, I don't think change time is good, we should probably use birth time, because when "it is created" is "built time on" date we are looking for I believe; can someone send a PR for this?

**Comment by adtya** [comment]
Yup, birth time makes more sense.

I can work on a PR for this, though I might need some help with testing the changes etc.

**Comment by RaitoBezarius** [comment]
> Yup, birth time makes more sense.
> 
> I can work on a PR for this, though I might need some help with testing the changes etc.

No problem. :)

**Comment by nikstur** [comment]
This reminds me of #95 which was fixed in #100. What version of Lanzaboote are you using?

**Comment by adtya** [comment]
I've been using lanzaboote/master in my flake.nix. The bootctl output I've posted above is of a generation added with lanzaboote/master.

With the changes from #139 I have verified that the date is being populated correctly.
```
Default Boot Loader Entry:
         type: Boot Loader Specification Type #2 (.efi)
        title: NixOS Stoat 23.05.20230319.60c1d71 (Linux 6.2.6-zen1) (Generation 119, Built on 2023-03-23)
           id: nixos-generation-119.efi
       source: /boot/EFI/Linux/nixos-generation-119.efi
     sort-key: lanza
      version: Generation 119, Built on 2023-03-23
        linux: EFI/Linux/nixos-generation-119.efi
      options: init=/nix/store/fj9v6z17mb310d2xq9bjm548lrhrd5a6-nixos-system-Skipper-23.05.20230319.60c1d71/init quiet
```

It seems that `fs::metadata(path)` used in #100 will traverse through symlinks and read the metadata from the destination file itself.

**Comment by RaitoBezarius** [comment]
> This reminds me of #95 which was fixed in #100. What version of Lanzaboote are you using?

On master, I am experiencing this bug too.

**Comment by adtya** [comment]
closing this as the issue has been fixed by #139
